src/replication.c

```diff
@@ -2386,6 +2481,353 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        return;
+
+    write_error: /* Handle sendCommand() errors. */
+        serverLog(LL_WARNING, "Sending command to primary in rdb channel replication handshake: %s", err);
+        sdsfree(err);
+        goto error;
+}
+
+
+/* Initialize server.pending_repl_data infrastructure, we will allocate the buffer itself once we need it */
+void replDataBufInit(void) {
+    serverAssert(server.pending_repl_data.blocks == NULL);
+    server.pending_repl_data.len = 0;
+    server.pending_repl_data.blocks = listCreate();
+    server.pending_repl_data.blocks->free = zfree;
+}
+
+/* Track replication data streaming progress, and serve clients from time to time */
+void replStreamProgressCallback(long long offset, int readlen) {
+    if (server.loading_process_events_interval_bytes &&
+        (offset + readlen)/server.loading_process_events_interval_bytes > offset/server.loading_process_events_interval_bytes)
+    {
+        replicationSendNewlineToMaster();
+        processEventsWhileBlocked();
+    }
+    atomicIncr(server.stat_repl_processed_bytes, readlen);
+}
+
+/* Reads replication data from primary into specified repl buffer block */
+int readIntoReplDataBlock(connection *conn, replDataBufBlock *o,  size_t read) {
+    int nread = connRead(conn, o->buf + o->used, read);
+    if (nread == -1) {
+        if (connGetState(conn) != CONN_STATE_CONNECTED) {
+            serverLog(LL_VERBOSE, "Error reading from primary: %s",connGetLastError(conn));
+            abortRdbConnectionSync(1);
+        }
+        return C_ERR;
+    }
+    if (nread == 0) {
+        if (server.verbosity <= LL_VERBOSE) {
+            sds info = catClientInfoString(sdsempty(), connGetPrivateData(conn));
+            serverLog(LL_VERBOSE, "Client closed connection %s", info);
+            sdsfree(info);
+        }
+        abortRdbConnectionSync(1);
+        return C_ERR;
+    }
+    o->used += nread;
+    incrReadsProcessed(nread);
+    return read - nread;
+}
+
+int isReplicaBufferLimitReached(void) {
+    return server.pending_repl_data.len > server.client_obuf_limits[1].hard_limit_bytes;
+}
+
+/* Read handler for buffering incoming repl data during RDB download/loading. */
+void bufferReplData(connection *conn) {
+    client *c = connGetPrivateData(conn);
+    size_t readlen = PROTO_IOBUF_LEN;
+    int read = 0;
+    
+    while (readlen > 0) {
+        listNode *ln = listLast(server.pending_repl_data.blocks);
+        replDataBufBlock *tail = ln ? listNodeValue(ln) : NULL;
+
+        /* Append to tail string when possible */
+        if (tail && tail->used < tail->size) {
+            size_t avail = tail->size - tail->used;
+            read = min(readlen, avail);
+            readlen -= read;
+            read = readIntoReplDataBlock(conn, tail, read);
+        }
+        if (readlen && read == 0) {
+            if (isReplicaBufferLimitReached()) {
+                serverLog(LL_DEBUG, "Replication buffer limit reached, stopping buffering");
```

Comment:
I agree. Upon reaching the limit, we will set the read handler to null. 
*also this log should be at the NOTICE level.

Reply:
Setting buffer limits poses another problem. Due to the fact that the replica is unaware of the total size of the snapshot, it may read too much repl-data and never be able to load the snapshot. In order to determine the buffer limits, it would be useful to share the expected snapshot size with the replica.

Reply:
we don't know the expected snapshot size.
i think a notice level print, and a suspension of the read handler should be enough.
then the buffer will accumulate on the master's side, and eventually lead to disconnection due to slave buffer limit?
p.s. we'll have to have a test that covers these paths.

Reply:
As a snapshot size assessment, we can look at the master's `used_memory_dataset`

Reply:
this is very inaccurate. it can easily be 2 time higher or lower (or even 10 times in pathological cases).
i'm not sure i understand what exactly you wanna do with that info. maybe that'll still be valid in some way.

Reply:
Ideally, I would like to prevent a case in which the replica side buffer accumulates too much data, and the replica is unable to load the RDB. When that happens, I would like the master's COB to take a part or to fail the sync operation. I don't want the sync to continue if it has no chance of succeeding.

Reply:
if the sync won't continue, all we can do is drop the connection and retry (hoping the conditions change, e.g. more memory will be available later).

you can't afford false negatives (unnecessary dropping of the connection if it could succeed), and if there are also false positives (not detecting correctly that it's hopeless), then i wonder if we should bother to implement anything based on the estimated rdb file size.
let's keep it simple, read as much as we can, let the rest be accumulated in the master, until it disconnects us.